### ISSUE TITLE
How to resolve the semantic versioning vs apple-style release branches problem?

### ISSUE THREAD

**Issue by icanzilb** [icanzilb/MarkCodable#15]
I must say I haven't looked too deeply into this but it seems that because of `swift-markdown` being a dependency `MarkCodable` cannot use semantic versioning.

`swift-markdown` uses branches for versioning: https://github.com/apple/swift-markdown/branches and SwiftPM doesn't allow `MarkCodable` to depend on a branch or even a commit-hash when I tried to use semantic versions.

For now I've created a `release/swift-5.7` branch for `MarkCodable` but ideally I'd use semantic versioning.

Note: I can't currently add `MarkCodable` to https://swiftpackageindex.com/add-a-package because it doesn't have a semantic versioned release tag 🤷🏽‍♂️

**Comment by shaps80** [comment]
Yeah I have a library that I wanted to open source that also depends on swift-markdown, came across the same issue. Perhaps I'll release it under a branch like you did, I was holding off a bit, hopeful for a semantically versioned release soonish, but its been months now :(

**Comment by icanzilb** [comment]
> that also depends on swift-markdown, came across the same issue

@shaps80 have you maybe posted in the Swift forums? There must be some kind of strategy for packages like ours?

**Comment by shaps80** [comment]
I haven't no. As far I understood, packaging against a branch only, is essentially pre-release.

In fact iPad Playgrounds can't even use non-semantically-versioned packages as a result, which is actually where I first noticed the issue.

I was a part of the Playgrounds beta's, so I did raise an issue with that there. But not specifically in the forums, might need to 👍

**Comment by icanzilb** [comment]
It seems the common theme is to fork & semver on your own: https://forums.swift.org/t/whats-the-recommended-way-to-depend-on-apple-packages-with-non-semantic-versioning/60125/11 

Which I might as well do if my choice is between:

 1. Maintain a legacy 5.6 branch and duplicate the comit tree over there + force any users to forfeit semantic versioning for their packages
 2. Make a fork a tag a new release 2-3 times a year 

I can see how either of these will hurt `MarkCodable` but I didn't find a better way yet.

**Comment by shaps80** [comment]
Nice one! I'll take look. I thought I'd already tried this but had issues with some of the C deps in that package, I might be mis-remembering though. Thanks for posting the update 👍

**Comment by DivineDominion** [comment]
FWIW there's also still some support for mirroring repositories on GitHub:
https://docs.github.com/en/repositories/creating-and-managing-repositories/duplicating-a-repository

I've tested the approaches listed there and they're virtually identical. 

I looked for this because there's a GH mirror of the GNU Emacs code repo here: https://github.com/emacs-mirror/emacs -- but that's rather related to importing stuff from another VCS

Automatically picking up changed from upstream to keep the mirror current would be really nice, but that requires a cron job somewhere. There's tools like https://github.com/beefsack/git-mirror to make this simpler, but meh.

Also, `git push --mirror` overrides all changes of one's copy. In your situation, your release tags will be removed. That sucks.

So best thing we have is fetching from apple/swift-markdown and push to icanzilb/swift-markdown, i.e. maintaining a fork, not a mirror. Sheesh.

(Somewhat related: Maybe it could make sense to create a MarkCodable GH org to make discovery of related packages easier than on your personal account @icanzilb)

**Comment by icanzilb** [comment]
> (Somewhat related: Maybe it could make sense to create a MarkCodable GH org to make discovery of related packages easier than on your personal account @icanzilb)

Agreed, I'm still trying to gauge what the reaction to this package is but if there's gonna be dependency forks flying around as well, there's definitely a necessity to create a separate org. I've still not lost hope there might be a different solution though, the last thing I want is to maintain forks of apple's repos.

**Comment by shaps80** [comment]
Yep, I'm in the same boat tbh. But, I really want to share this library as its pretty cool I think and I imagine a lot of projects could use it. I'd like to make it available so that's my plan this week ;)
